### ISSUE TITLE
error collecting concat of empty joins with Streaming API

### ISSUE THREAD

**Issue by KDruzhkin** [pola-rs/polars#10971]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

data_1 = pl.DataFrame({"A": [1], "B": [11]}).lazy()
data_2 = pl.DataFrame({"A": [2], "B": [22]}).lazy()

query = pl.DataFrame({"A": [3]}).lazy()

empty_join_1 = query.join(data_1, on="A")
empty_join_2 = query.join(data_2, on="A")

empty_concat = pl.concat([empty_join_1, empty_join_2])

empty_concat.collect(streaming=False)
# => OK

empty_concat.collect(streaming=True)
# => thread '<unnamed>' panicked at .../polars/polars/crates/polars-pipe/src/executors/sinks/ordered.rs:41:66:
# called `Option::unwrap()` on a `None` value
```

### Log output

```shell
UNION: `parallel=false` union is run sequentially
CACHE SET: cache id: 57dca74215084b8a
CACHE HIT: cache id: 57dca74215084b8a
join parallel: false
inner join: keys are sorted: use sorted merge join
INNER join dataframes finished
join parallel: false
inner join: keys are sorted: use sorted merge join
INNER join dataframes finished
RUN STREAMING PIPELINE
df -> generic_join_build
RefCell { value: [df -> placeholder -> pass -> ordered_sink, df -> generic_join_build, df -> placeholder -> pass -> ordered_sink] }
thread '<unnamed>' panicked at /home/runner/work/polars/polars/crates/polars-pipe/src/executors/sinks/ordered.rs:41:66:
called `Option::unwrap()` on a `None` value
```

### Issue description

Some lazy queries execute normally with `streaming=False`, but fail with `streaming=True`. As far as I can see, failure happens when there are empty intermediate results.

### Expected behavior

Queries should return the same results in streaming and non-streaming modes.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.2
Index type:          UInt32
Platform:            Linux-6.2.0-31-generic-x86_64-with-glibc2.35
Python:              3.9.18 (main, Aug 25 2023, 13:20:14) 
[GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          0.3.1
deltalake:           <not installed>
fsspec:              2023.9.0
matplotlib:          <not installed>
numpy:               1.25.2
pandas:              2.1.0
pyarrow:             13.0.0
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by reswqa** [comment]
We have four pipelines in order:
1.  `df -> generic_join_build`
2. `df -> placeholder -> pass -> ordered_sink`
3. `df -> generic_join_build`
4.  `df -> placeholder -> pass -> ordered_sink`.

To perform `union`, the `reduce_sink` generated by `2` and `4` should be combined as they shared the same sink. But what we actually combined was the sink generated by `3`(i.e. `GenericBuild`). This does seem like a problem.

**Comment by KDruzhkin** [comment]
The bug persists in polars **0.20.2**.

**Comment by AroneyS** [comment]
Still fails in v0.20.3

**Comment by KDruzhkin** [comment]
Fixed in 0.20.9.
